modules/ingester/instance.go

```diff
@@ -376,30 +376,36 @@ func (i *instance) GetBlockToBeFlushed(blockID uuid.UUID) *LocalBlock {
 	return nil
 }
 
-func (i *instance) ClearFlushedBlocks(completeBlockTimeout time.Duration) error {
-	var err error
-
+func (i *instance) ClearOldBlocks(flushObjectStorage bool, completeBlockTimeout time.Duration) error {
 	i.blocksMtx.Lock()
 	defer i.blocksMtx.Unlock()
 
-	for idx, b := range i.completeBlocks {
-		flushedTime := b.FlushedTime()
-		if flushedTime.IsZero() {
+	cutoff := time.Now().Add(-completeBlockTimeout)
+	for idx, max := 0, len(i.completeBlocks); idx < max; idx++ {
+		b := i.completeBlocks[idx]
+
+		// Keep blocks awaiting flush if we are flushing to object storage.
+		if flushObjectStorage && b.FlushedTime().IsZero() {
 			continue
 		}
 
-		if flushedTime.Add(completeBlockTimeout).Before(time.Now()) {
-			i.completeBlocks = append(i.completeBlocks[:idx], i.completeBlocks[idx+1:]...)
+		// Keep blocks within retention.
+		if b.BlockMeta().EndTime.After(cutoff) {
+			continue
+		}
 
-			err = i.local.ClearBlock((uuid.UUID)(b.BlockMeta().BlockID), i.instanceID)
-			if err == nil {
-				metricBlocksClearedTotal.Inc()
-			}
-			break
+		// This block can be deleted.
+		err := i.local.ClearBlock((uuid.UUID)(b.BlockMeta().BlockID), i.instanceID)
+		if err != nil {
+			return err
```

Comment:
Does this mean a bad block will prevent any flush from ever happening?

Reply:
True, if it failed repeatedly it would mean that no blocks later in the slice would be deleted off the local disk, if the sliced maintained the same order (I think it does).  Flushing to object storage and persisting data off the ingester wouldn't be affected.  Can say that I've never seen this happen in 4 years of operations, so I'm ok with this as-is.

Reply:
Sounds reasonable!